scripts/mbedtls_framework/test_case.py

```diff
@@ -89,3 +111,28 @@ def write_data_file(filename: str,
             tc.write(out)
         out.write('\n# End of automatically generated file.\n')
     os.replace(tempfile, filename)
+
+def psa_or_3_6_feature_macro(psa_alg: str,
+                             domain_3_6: str) -> str:
+    """Determine the dependency symbol for a given psa_alg based on
+       the domain and Mbed TLS version. For more information about the domains,
+       and MBEDTLS_MD_CAN_ prefixed symbols, see transition-guards.md.
+    """
+
+    if domain_3_6 == "DOMAIN_3_6_PSA" or not build_tree.is_mbedtls_3_6():
+        if psa_alg in PK_MACROS_3_6 or psa_alg in HASHES_3_6:
+            return psa_alg
+        if psa_alg.startswith('PSA_ALG_') and \
+            psa_alg[8:] in ['MD5', 'RIPEMD160', 'SHA_1', 'SHA_224', 'SHA_256',
+                            'SHA_384', 'SHA_512', 'SHA3_224',
+                            'SHA3_256', 'SHA3_384', 'SHA3_512']:
+            return psa_information.psa_want_symbol(psa_alg)
+
+    if psa_alg in HASHES_3_6:
+        return HASHES_3_6[psa_alg]
+    if psa_information.psa_want_symbol(psa_alg) in HASHES_3_6:
+        return HASHES_3_6[psa_information.psa_want_symbol(psa_alg)]
+    if psa_alg in PK_MACROS_3_6:
```

Comment:
Perhaps add "and domain is `USE_PSA`" - otherwise it means the user did something wrong. (Then we'll fall through to the `raise ValueError` below.)

Reply:
The domains are still a bit confusing to me: from what I understand in 3.6 USE_PSA domain means we might use both PSA and legacy symbols depending on whether USE_PSA is enabled or not. But this function does not really reflect that with PK symbols.

Reply:
Yes, things aren't very neat here. I think two things are going on:
1. The `psa_alg` and `domain` argument are not fully independent: all users of PK (and PK itself) are in the `USE_PSA` domain. Contrast with say AES: some users are in the legacy domain (for example PKCS5), some are in the `USE_PSA` domain (for example TLS), some in the PSA domain (PBKDF-AES-CMAC), so they're each going to use different macros for "is AES available".
2. Things in the `USE_PSA` domain, like PK, tend to define custom macros for "is XXX available" that take into account whether `USE_PSA` is enabled or not - so users don't have to be like "is `USE_PSA` on? then I'll use macro XXX, else I'll use macro YYY", then can just use the abstracted macro, like `MBEDTLS_PK_HAVE_ECC_KEYS` which expresses "PK has support for ECC keys (and you don't need to care whether it's based on PSA or legacy)".

I hope it makes sense. To be honest, this whole situation is really not an example of a clean architecture. From a purely engineering point of view, the right thing would have been to first deliver 4.0 where everything is based on PSA, and then work on driver-only builds. But we had to deliver driver-only earlier, so we did what we had to do, but the result isn't exactly pretty.